coursefunctions.h

```diff
@@ -390,41 +390,66 @@ void TestVertexShader(Buffer2D<PIXEL> & target)
 
         PIXEL colors[3] = {0xffff0000, 0xff00ff00, 0xff0000ff};
         // Your code for 'colorAttributes' goes here
+        colorAttributes[0].attrs[0] = 1.0;
+	colorAttributes[0].attrs[1] = 0.0;
+        colorAttributes[0].attrs[2] = 0.0;
+
+	colorAttributes[1].attrs[0] = 0.0;
+        colorAttributes[1].attrs[1] = 1.0;
+	colorAttributes[1].attrs[2] = 0.0;
+	
+	colorAttributes[2].attrs[0] = 0.0;
+        colorAttributes[2].attrs[1] = 0.0;
+	colorAttributes[2].attrs[2] = 1.0;
 
         FragmentShader myColorFragShader;
+        myColorFragShader.FragShader = ColorFragShader;
 
         Attributes colorUniforms;
         // Your code for the uniform goes here, if any (don't pass NULL here)
         
         VertexShader myColorVertexShader;
-        // Your code for the vertex shader goes here 
+        // Your code for the vertex shader goes here
+        myColorVertexShader.VertShader = vertexShader; 
 
         /******************************************************************
-		 * TRANSLATE (move +100 in the X direction, +50 in the Y direction)
-         *****************************************************************/
+	 * TRANSLATE (move +100 in the X direction, +50 in the Y direction)
+        *****************************************************************/
         // Your translating code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
+        Matrix transMatrix(3, 3);
+        transMatrix.translation(100, 50, 0);
+        colorUniforms.matrixAttributes = transMatrix;
 
 	DrawPrimitive(TRIANGLE, target, colorTriangle, colorAttributes, &colorUniforms, &myColorFragShader, &myColorVertexShader);
 
         /***********************************
          * SCALE (scale by a factor of 0.5)
          ***********************************/
         // Your scaling code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
+        Matrix scaleMatrix(3, 3);
+        scaleMatrix.scaling(0.5, 0.5, 1);
+        colorUniforms.matrixAttributes = scaleMatrix;
 
         DrawPrimitive(TRIANGLE, target, colorTriangle, colorAttributes, &colorUniforms, &myColorFragShader, &myColorVertexShader);
 
         /**********************************************
          * ROTATE 45 degrees in the X-Y plane around Z
          *********************************************/
         // Your rotation code that integrates with 'colorUniforms', used by 'myColorVertexShader' goes here
+        Matrix rotMatrix(3, 3);
+        rotMatrix.rotation(M_PI/4);
```

Comment:
You could have your rotation function create a matrix for you inside of itself which would eliminate the need for line 439.